Improve typing of QueryBuilder.all()

danielhollas opened the pull request:
Provide an [overloaded method definitions](https://typing.python.org/en/latest/spec/overload.html#overload-definitions) for `flat=True` and `flat=False` arguments. 
This allows us to get rid of multiple `type: ignore` comments.

danielhollas commented on src/aiida/orm/querybuilder.py:
```diff
@@ -1105,7 +1105,13 @@ def iterdict(self, batch_size: Optional[int] = 100) -> Iterable[Dict[str, Dict[s
             yield item
 
-    def all(self, batch_size: Optional[int] = None, flat: bool = False) -> Union[List[List[Any]], List[Any]]:
+    @overload
```
Note that the same approach is already used for the `first()` method above

khsrali commented on src/aiida/orm/querybuilder.py:
```diff
@@ -1105,7 +1105,13 @@ def iterdict(self, batch_size: Optional[int] = 100) -> Iterable[Dict[str, Dict[s
 
-    def all(self, batch_size: Optional[int] = None, flat: bool = False) -> Union[List[List[Any]], List[Any]]:
+    @overload
+    def all(self, batch_size: int | None = None, flat: Literal[False] = False) -> list[list[Any]]: ...
```
I'm not sure, but I think the default value here is redundant 
```suggestion
    def all(self, batch_size: int | None = None, flat: Literal[False]) -> list[list[Any]]: ...
```

danielhollas replied:
Unfortunately, without a default  value I get a syntax error because the previous argument has a default value.

danielhollas replied:
Yeah, that would make sense, but unfortunately mypy complains

> src/aiida/orm/querybuilder.py:1112: error: Incompatible default for argument "flat" (default has type "Literal[False]", argument has type "Literal[True]")  [assignment]

I am not sure if there's a different solution here, I don't have experience with `@overload` decorator. Maybe @unkcpz or @sphuber have ideas?

danielhollas commented:
I've snuck in a mypy update (1.16 to 1.17), no typing changes needed.

GeigerJ2 commented:
Thanks, @danielhollas! This typing problem has been bothering my every time I was using QB. Thanks for taking the time to fix it properly. Will review during the coding days!

GeigerJ2 reviewed (commented):
Nice, thanks a lot! Want to also remove related `typing.cast` calls? Bc I had added a few of those :D
E.g., here: https://github.com/danielhollas/aiida-core/blob/87f70904ee6ef17778c810a859b49169cee3078f/src/aiida/tools/_dumping/executors/profile.py#L94

danielhollas commented on src/aiida/tools/archive/create.py:
```diff
@@ -16,7 +16,7 @@
 from datetime import datetime
 from pathlib import Path
-from typing import Callable, Dict, Iterable, List, Optional, Sequence, Set, Tuple, Union
+from typing import Callable, Iterable, Optional, Sequence, Union
```
Yes, sorry about this, I'll revert it.

danielhollas replied:
Ugh, I take that back, it would be too annoying to revert it now :sweat_smile: Instead, I promise to not do that again :D My general disposition as a PR author is to do little cleanups sometimes in files that I touch, but as a reviewer I hate that :D

I don't even think its worth the churn to change this globally, the old syntax is "fine".

GeigerJ2 replied:
All valid points hehe :D

danielhollas commented:
> Want to also remove related typing.cast calls? Bc I had added a few of those :D

Oh, nice, I would have never find those. I've removed them.
(I honestly am not sure if it is better to use cast or type: ignore comments, need to read up on that).

Thanks for the review!

GeigerJ2 commented:
> Oh, nice, I would have never find those. I've removed them. (I honestly am not sure if it is better to use cast or type: ignore comments, need to read up on that).
> 
> Thanks for the review!

Both are just masking improper typing, so, ideally, we shouldn't have to add either 😅
